src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1068 @@
+    return Type::WidenMin;
+  }
+  if (_srange._lo == std::numeric_limits<S>::min() && _srange._hi == std::numeric_limits<S>::max() &&
+      _urange._lo == std::numeric_limits<U>::min() && _urange._hi == std::numeric_limits<U>::max() &&
+      _bits._zeros == U(0) && _bits._ones == U(0)) {
+    // bottom type
+    return Type::WidenMax;
+  }
+  return widen;
+}
+
+#ifdef ASSERT
+template <class S, class U>
+bool TypeIntPrototype<S, U>::contains(S v) const {
+  U u(v);
+  return v >= _srange._lo && v <= _srange._hi &&
+         u >= _urange._lo && u <= _urange._hi &&
+         _bits.is_satisfied_by(u);
+}
+
+// Verify that this set representation is canonical
+template <class S, class U>
+void TypeIntPrototype<S, U>::verify_constraints() const {
+  // Assert that the bounds cannot be further tightened
+  assert(contains(_srange._lo) && contains(_srange._hi) &&
+         contains(S(_urange._lo)) && contains(S(_urange._hi)), "");
+
+  // Assert that the bits cannot be further tightened
+  if (U(_srange._lo) == _urange._lo) {
+    assert(!adjust_bits_from_unsigned_bounds(_bits, _urange)._progress, "");
+  } else {
+    RangeInt<U> neg_range{U(_srange._lo), _urange._hi};
+    auto neg_bits = adjust_bits_from_unsigned_bounds(_bits, neg_range);
+    assert(neg_bits._present, "");
+    assert(!adjust_unsigned_bounds_from_bits(neg_range, neg_bits._result)._progress, "");
+
+    RangeInt<U> pos_range{_urange._lo, U(_srange._hi)};
+    auto pos_bits = adjust_bits_from_unsigned_bounds(_bits, pos_range);
+    assert(pos_bits._present, "");
+    assert(!adjust_unsigned_bounds_from_bits(pos_range, pos_bits._result)._progress, "");
+
+    assert((neg_bits._result._zeros & pos_bits._result._zeros) == _bits._zeros &&
+           (neg_bits._result._ones & pos_bits._result._ones) == _bits._ones, "");
+  }
+}
+#endif // ASSERT
+
+template class TypeIntPrototype<jint, juint>;
+template class TypeIntPrototype<jlong, julong>;
+template class TypeIntPrototype<intn_t<1>, uintn_t<1>>;
+template class TypeIntPrototype<intn_t<2>, uintn_t<2>>;
+template class TypeIntPrototype<intn_t<3>, uintn_t<3>>;
+template class TypeIntPrototype<intn_t<4>, uintn_t<4>>;
+
+// Compute the meet of 2 types. When dual is true, the subset relation in CT is
+// reversed. This means that the result of 2 CTs would be the intersection of
+// them if dual is true, and be the union of them if dual is false. The subset
+// relation in the Type hierarchy is still the same, however. E.g. the result
+// of 1 CT and Type::BOTTOM would always be Type::BOTTOM, and the result of 1
+// CT and Type::TOP would always be the CT instance itself.
+template <class CT, class S, class U>
+const Type* TypeIntHelper::int_type_xmeet(const CT* i1, const Type* t2, make_type_t<S, U> make, bool dual) {
+  // Perform a fast test for common case; meeting the same types together.
+  if (i1 == t2 || t2 == Type::TOP) {
+    return i1;
+  }
+  const CT* i2 = t2->try_cast<CT>();
```

Comment:
And what is the significance of the `dual` state of i1 and `i2`? Should their `dual` state be the same?
Can we add some assert here?

Reply:
I decide to make `TypeIntHelper` a friend of `TypeInt` and use `i1->_is_dual` here instead. I think it makes it clearer.